src/ar/ARTester.cpp

```diff
@@ -1,105 +1,326 @@
 #include <opencv2/imgproc.hpp>
 
-#include <iostream>
+#include "../ThreadedCapture.h"
+#include "Detector.h"
+
+// change to true to turn on timing code
+#define TIMING false
 
 const std::string ORIG_WINDOW_NAME = "Image (Original)";
-const std::string THRESH_TRACKBAR_NAME = "Threshold";
-const std::string THRESH2_TRACKBAR_NAME = "Threshold 2";
-const std::string BLUR_TRACKBAR_NAME = "Blur";
 
 constexpr bool EXTRA_WINDOWS = true;
 
+// Set to whichever camera params should be used
+AR::CameraParams PARAMS = AR::getCameraParams(AR::Params::ROBOT_TOP_WEBCAM_480);
+
+// Set to whichever MarkerSet should be used (CIRC/URC)
+std::shared_ptr<AR::MarkerSet> MARKER_SET =
+	AR::Markers::CIRC_MARKERS(); // AR::Markers::URC_MARKERS();
+
+const std::string keys = "{h help || Show this help message.}"
+						 "{params p       | 0 | The set of camera parameters to use. This "
+						 "value should be a non-negative zero-indexed integer corresponding "
+						 "to the member of AR::Params you want to use.}"
+						 "{marker_set m   | urc | The set of markers to look for. Currently "
+						 "only \"urc\" and \"circ\" are supported.}"
+						 "{cam c | 0 | The camera ID to use.}"
+						 "{frame_by_frame f || If present, program will go frame-by-frame "
+						 "instead of capturing continuously.}";
+
 int camera_id = 0;
 
+std::vector<cv::Point2d> projectCube(float len, cv::Vec3d rvec, cv::Vec3d tvec)
+{
+	std::vector<cv::Point3d> object_points;
+	std::vector<cv::Point2d> image_points;
+
+	object_points.push_back(cv::Point3d(-(len / 2), (len / 2), 0));
+	object_points.push_back(cv::Point3d((len / 2), (len / 2), 0));
+	object_points.push_back(cv::Point3d((len / 2), -(len / 2), 0));
+	object_points.push_back(cv::Point3d(-(len / 2), -(len / 2), 0));
+	object_points.push_back(cv::Point3d(-(len / 2), (len / 2), len));
+	object_points.push_back(cv::Point3d((len / 2), (len / 2), len));
+	object_points.push_back(cv::Point3d((len / 2), -(len / 2), len));
+	object_points.push_back(cv::Point3d(-(len / 2), -(len / 2), len));
+	cv::projectPoints(object_points, rvec, tvec, PARAMS.getCameraParams(),
+					  PARAMS.getDistCoeff(), image_points);
+
+	return image_points;
+}
+
+std::vector<cv::Point2f> projectGrid(cv::Size imageSize, int spacing)
+{
+	cv::Point2f center(imageSize.width / 2, imageSize.height / 2);
+	std::vector<cv::Point2f> grid_points;
+	std::vector<cv::Point2f> projected_points;
+	for (int x = 0; x < imageSize.width / 2; x += spacing)
+	{
+		for (int y = 0; y < imageSize.height / 2; y += spacing)
+		{
+			grid_points.push_back(cv::Point2f(x, y) + center);
+			if (x != 0 || y != 0)
+			{
+				grid_points.push_back(cv::Point2f(-x, -y) + center);
+				grid_points.push_back(cv::Point2f(-x, y) + center);
+				grid_points.push_back(cv::Point2f(x, -y) + center);
+			}
+		}
+	}
+	cv::undistortPoints(grid_points, projected_points, PARAMS.getCameraParams(),
+						PARAMS.getDistCoeff());
+	return projected_points;
+}
+
 int main(int argc, char *argv[])
```

Comment:
This is a REALLY long function. In my experience, it's easier to maintain code when it is broken into multiple smaller functions. We don't have to fix that for this PR, but it's something to think about for the future.

Reply:
Yeah, agreed. I can still fix it for the PR or later, it would be good to have it be more maintainable in the future